Add function to wrap text, wrap option help strings to 77 chars

nilsdeppe opened the pull request:
## Proposed changes

- Add a function to wrap text with indentation to a specified number of characters
- Have options automatically wrap help string output to 77 characters.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Add function wrap_text

wrap_text wraps long lines of text so that they can be nicely printed to the
terminal.

wthrowe commented on src/Utilities/WrapText.hpp:
```diff
@@ -0,0 +1,15 @@
+/// \brief Wrap the string `str` so that it is no longer than `line_length` and
+/// indent each new line with `indentation`.
+///
+/// Single words longer than `line_length` are hyphenated.
```
Document that the first line is not indented.  (Or maybe change it to indent the line, since both use cases indent that line themselves.)

nilsdeppe replied:
Good idea! I did this and pushed fixups as well as rebased on develop

wthrowe commented:
Looks good.  Squash.

nilsdeppe commented:
Squashed. Thanks for the review @wthrowe!

moxcodes commented on src/Utilities/WrapText.cpp:
```diff
@@ -0,0 +1,44 @@
+    const auto last_newline =
+        str.rfind('\n', i + line_length - indentation.size());
+    if (last_newline != std::string::npos and
+        last_newline > i + 1 + indentation.size()) {
```
I'm a little confused about the indentation logic. So, if the string has two newlines with a gap of less than the indentation, the condition on this if will end up false -- when the string in the test is `..."wrap \nAA\nthis text"...` and the indentation length 3, it misses indenting the "AA"

moxcodes replied:
Did you fix this edge case? I think it's not in the collection of example outputs you posted.

nilsdeppe replied:
I believe the last test case covers this:
```
\n\nthis text\nand this text\n
```

moxcodes replied:
but the `this text` is not shorter than the indentation length -- my concern is that if the newlines are spaced less than the indentation length, followed by a long span of characters, this if seems to skip adding the indentation after one of them.

nilsdeppe replied:
Ah, sorry I missed that. Okay, I'll add a check for that. Is the following able to correctly check what you want?
```
void foog3() {
  const size_t length = 50;
  const auto result = wrap_text(
      "I want to wrap \n\na\nand this text\n"
      "thisisareallylongwordthatwouldcauseproblemssoweneedtoerror "
      "after about 20 characters please.",
      length, "    ");
  std::cout << "foog3\n";
  stream_dashes(length);
  std::cout << result << "\n";
  stream_dashes(length);
}

```
which results in:
```
foog3
--------------------------------------------------
    I want to wrap

    a
    and this text
    thisisareallylongwordthatwouldcauseproblemsso-
    weneedtoerror after about 20 characters
    please.
--------------------------------------------------
```

moxcodes replied:
Yep, I think that satisfies the remaining case. Thanks!

moxcodes commented on src/Utilities/WrapText.cpp:
```diff
@@ -0,0 +1,44 @@
+
+    const auto last_space =
+        str.rfind(' ', i + line_length - 1 - indentation.size());
+    if (last_space < i + 1 + indentation.size() or
```
This conditional seems not quite right also, consider a string that just has an extremely large block of characters.
`wrap_text("AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
"AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
"AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
"AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
50, "          ")`
gives me some lines of length 58.

moxcodes replied:
note: that " " is supposed to have 10 spaces in it but github is too clever and strips most of the spaces -.-

moxcodes reviewed (changes requested):
I think I noticed some edge cases, or at least some conditions for which I expected this utility to act differently from how it actually behaves.

nilsdeppe commented:
I rebased on develop and updated the algorithm to handle the edge cases @moxcodes found. I updated the tests. Here is the output of the test wrapping where the dashes is the number of characters limited to:
```
fooa
--------------------------------------------------
         aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa-
         aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa-
         aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa-
         aaaaaaaaaaaaaaaaaa
--------------------------------------------------
fooa2
--------------------------------------------------
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa-
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa-
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
--------------------------------------------------
fooa3
--------------------------------------------------
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa-
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
--------------------------------------------------
foob
--------------------------------------------------
   thisisareallylongwordthatwouldcauseproblemssow-
   eneedtoerror I want to wrap

   this text
   thisisareallylongwordthatwouldcauseproblemssow-
   eneedtoerror after about 50 characters please.
--------------------------------------------------
fooc
--------------------------------------------------
thisisareallylongwordthatwouldcauseproblemssowene-
edtoerror I want to wrap

this text
thisisareallylongwordthatwouldcauseproblemssowene-
edtoerror after about 50 characters please.
--------------------------------------------------
food
--------------------------------------------------
   thisisareallylongwordthatwouldcauseproblemssow-
eneedtoerror I want to wrap

this text
thisisareallylongwordthatwouldcauseproblemssowene-
edtoerror after about 50 characters please.
--------------------------------------------------
fooe
--------------------------------------------------
I want to wrap

this text
thisisareallylongwordthatwouldcauseproblemssowene-
edtoerror after about 50 characters please.
--------------------------------------------------
foof
--------------------
I want to wrap

this text
thisisareallylongwo-
rdthatwouldcausepro-
blemssoweneedtoerror
after about 20
characters please.
--------------------
foog
--------------------
I want to wrap

this text
thisisareallylongwo-
rdthatwouldcausepro-
blemssoweneedtoerror
after about 20
characters please.
--------------------
foog2
--------------------------------------------------
    I want to wrap

    this text
    and this text
    thisisareallylongwordthatwouldcauseproblemsso-
    weneedtoerror after about 20 characters
    please.
--------------------------------------------------
```

kidder commented:
@moxcodes @wthrowe please look at

nilsdeppe committed:
Make option printing wrap text to 77 chars

nilsdeppe commented:
Okay, rebased on develop and added the test case

moxcodes reviewed (approved):
Looks good to me
